core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -1143,14 +1156,20 @@ public static Expression convert(Expression operand, Type fromType,
                 "toString",
                 operand));
       } else {
-        // E.g. from "BigDecimal" to "String"
-        // Generate "x == null ? null : x.toString()"
-        return Expressions.condition(
-            Expressions.equal(operand, RexImpTable.NULL_EXPR),
-            RexImpTable.NULL_EXPR,
-            Expressions.call(
-                operand,
-                "toString"));
+        Expression result;
+        try {
+          // Try to call "toString()" method
+          // E.g. from "Integer" to "String"
+          // Generate "x == null ? null : x.toString()"
+          result = Expressions.condition(
+              Expressions.equal(operand, RexImpTable.NULL_EXPR),
+              RexImpTable.NULL_EXPR,
+              Expressions.call(operand, "toString"));
+        } catch (Exception e) {
+          // No "toString()" method, fall through to (String)x
```

Comment:
can this every happen for no `toString` method?

Reply:
Theoretically, a Class should contain `toString` method. However, there are some special cases in Calcite.
For example (_QuidemTest#winagg.iq_), if the operand is `SqlFunctions.lesser(String.class, String.class)`,  its type will be `java.lang.Comparable`, which has no `toString `method, leading to Exception in `Expressions.call(operand, "toString")`.
Because for `String.class`, the method matched is:
```
  public static <T extends Comparable<T>> T lesser(T b0, T b1) {
    return b0 == null || b0.compareTo(b1) > 0 ? b1 : b0;
  }
```

Reply:
OK, got that.

Reply:
Although the operand type class of `SqlFunctions.lesser(String.class, String.class)` does not implement `toString` method, the object should have implement the `toString` method, because the java Object already has that, did you mean the operand object is not a java Object ?

Reply:
Yes, the code `...lesser(..).toString` will be ok in runtime. It just throws the exception during codegen in `Expressions.call(operand, "toString"))` through reflection. `(String)lesser(...)` skips this step.

Reply:
Do you think we can make an effort for `Expressions.call(operand, "toString"))` to get the right `toString` method ? Even though it uses the Java reflection.

Reply:
I investigate this problem, it may be complicated to implement this feature from begin. However, relying on Guava's `TypeResolver`, we can easily infer a method's return type according to its argument types, as demonstrated below.
Is it necessary to open another JIRA for this feature?

```
public static Type inferRuntimeReturnType(final Method method,
      final List<Expression> arguments) {
    final Type returnType = method.getReturnType();
    final Type genericReturnType = method.getGenericReturnType();
    if (returnType == genericReturnType) {
      return returnType;
    }
    Type runtimeReturnType = genericReturnType;
    final Type[] parameterTypes = method.getParameterTypes();
    final Type[] genericParameterTypes = method.getGenericParameterTypes();
    for (int i = 0; i < parameterTypes.length; i++) {
      if (parameterTypes[i] != genericParameterTypes[i]) {
        final Type argType = arguments.get(i).getType();
        final TypeResolver resolver =
            new TypeResolver().where(genericParameterTypes[i], argType);
        runtimeReturnType = resolver.resolveType(runtimeReturnType);
      }
    }
    assert runtimeReturnType != genericReturnType;
    return runtimeReturnType;
  }
```

Reply:
I'm not sure how this code works, compared to this code, i'm more inclined with the try catch way, do we need to refine the specific exception instead of just `Exception` ?

Reply:
Ok, let me narrow it down to RuntimeException.
```
  public static MethodCallExpression call(Expression target, String methodName,
      Iterable<? extends Expression> arguments) {
    Method method;
    try {
      //noinspection unchecked
      method = Types.toClass(target.getType())
          .getMethod(methodName, Types.toClassArray(arguments));
    } catch (NoSuchMethodException e) {
      throw new RuntimeException("while resolving method '" + methodName
          + "' in class " + target.getType(), e);
    }
    return call(target, method, arguments);
  }
```